lib/service/rpmsg/rpc/rpmsg_rpc_server.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <metal/mutex.h>
+#include <metal/spinlock.h>
+#include <metal/utilities.h>
+#include <openamp/open_amp.h>
+#include <openamp/rpmsg_rpc_client_server.h>
+#include <string.h>
+#include <fcntl.h>
+
+#define raw_printf(format, ...) printf(format, ##__VA_ARGS__)
+#define LPRINTF(format, ...) raw_printf("Master> " format, ##__VA_ARGS__)
+#define LPERROR(format, ...) LPRINTF("ERROR: " format, ##__VA_ARGS__)
+
+int request_termination;
+int ept_deleted;
+int err_cnt;
+static struct rpmsg_rpc_services *table[TABLE_SIZE];
+
+void *copy_from_shbuf(void *dst, void *shbuf, int len)
+{
+	void *retdst = dst;
+
+	while (len && ((uintptr_t)shbuf % sizeof(int))) {
+		*(uint8_t *)dst = *(uint8_t *)shbuf;
+		dst++;
+		shbuf++;
+		len--;
+	}
+	while (len >= (int)sizeof(int)) {
+		*(unsigned int *)dst = *(unsigned int *)shbuf;
+		dst += sizeof(int);
+		shbuf += sizeof(int);
+		len -= sizeof(int);
+	}
+	while (len > 0) {
+		*(uint8_t *)dst = *(uint8_t *)shbuf;
+		dst++;
+		shbuf++;
+		len--;
+	}
+	return retdst;
+}
+
+int linux_rpmsg_rpc_server_init(int id, int (*cb_function)
+				(void *, struct rpmsg_endpoint *))
+{
+	struct rpmsg_rpc_services *service =
+	(struct rpmsg_rpc_services *)malloc(sizeof(struct rpmsg_rpc_services));
+
+	service->id = id;
+	service->respcb.cb_function = cb_function;
+
+	int hash_index = id % TABLE_SIZE;
+
+	while (table[hash_index]) {
+		++hash_index;
+		hash_index %= TABLE_SIZE;
+	}
+
+	table[hash_index] = service;
+
+	return hash_index;
+}
+
+struct rpmsg_rpc_services *find_service(int id)
+{
+	int hash_index = id % TABLE_SIZE;
+
+	while (table[hash_index]) {
+
+		if ((int)table[hash_index]->id == id)
+			return table[hash_index];
+
+		++hash_index;
+		hash_index %= TABLE_SIZE;
+	}
+
+	return NULL;
+}
+
+int linux_rpmsg_endpoint_server_cb(struct rpmsg_endpoint *ept, void *data,
+				   size_t len,
+				   uint32_t src, void *priv)
+{
+	unsigned char buf[MAX_BUF_LEN];
+	unsigned long int id;
+
+	(void)priv;
+	(void)src;
+
+	/* In case the shared memory is device memory
+	 * E.g. For now, we only use UIO device memory in Linux.
+	 */
+	if (len > MAX_BUF_LEN)
+		len = MAX_BUF_LEN;
+	copy_from_shbuf(buf, data, len);
+	id = *buf;
+	struct rpmsg_rpc_services *service = find_service(id);
+
+	if (service) {
+		service->respcb.cb_function(buf, ept);
+	} else {
+		LPRINTF("\nHandling remote procedure call errors:\r\n");
+		raw_printf("rpc id %ld\r\n", id);
+		err_cnt++;
```

Comment:
Suppress the err_cnt and concatenate the  raw_printf and LPRINTF print in one metal_log(METAL_LOG_ERROR,...)